File: python/slidingwindow.py
'''
SLIDING WINDOW TECHNIQUE
useful for when theres overlapping subarrays 

start with initial subarray, 
sum it up

move up a value
substart previous and add the last value in subarray

1st sum = sum
2nd sum = sum + element that went in - element that went out

    
'''


class Solution:
    def fixed_sliding_window(self, arr:list[int], k:int)-> list[int]:
        curr = sum(arr[:k])
        output = [curr]

        for i in range(1, len(arr)-k+1):
            curr = curr - arr[i-1]
            curr = curr + arr[i+k-1]
            output.append(curr)
        
        #returns the sum of all subarrays given the length of k
        return output



print(Solution.fixed_sliding_window(Solution, [3,5,3,4,5,2], 2))


# O(n) CONSTANT TIMEj
def slide_window(arr:list[int], k:int) -> list[int]:
    curr = sum(arr[:k]) # first sum
    out = [curr] #puts the first sum in the output list

    for i in range(len(arr)-k):
        # i until start index of the last group (to not overflow)

        #basically calculates the sum for the next window (starting from after current)
        curr -= arr[i]
        curr += arr[i+k]
        out.append(curr)
    return out

print(slide_window([3,5,3,4,5,2], 2))

File: cpp/slidingwindow.cpp
#include <iostream>
#include <vector>
#include <string>

/*
SLIDING WINDOW ALGORITHM

COMMON IN CODING

g++ slidingwindow.cpp -o out\slidingwindow
out\slidingwindow

g++ slidingwindow.cpp -o linout/slidingwindow
out/slidingwindow

*/

struct node{
  int data;
  node* next;  

};

class linked_list{
    node* head = nullptr;
    public:
    linked_list(){

    }
    
};


class Solution{
    public:
    static std::vector<int> slide_window(const std::vector<int>& arr, const int& k) {
        std::vector<int> output;

        int curr = 0;
        for(int i=0; i<k; ++i){ //gets the current (initial sum)
            curr += arr[i];
        }
        output.push_back(curr);
        for(int i =0; i<arr.size()-k; ++i) {
            output.push_back(curr - arr[i] + arr[i+k]);
        };
    
        return output;

    }
};

template<typename T>
std::ostream& operator<<(std::ostream& out, const std::vector<T>& temp) {
    int i;
    for(i=0; i<temp.size()-1; ++i) {
        out<< temp[i] << ' ';
    }
    out<<temp[i]<<'\n';
    return out;
}


int main() {
    std::cout<< Solution::slide_window({4,3,2,5,3,5},2);
}